### ISSUE TITLE
app crash on [DDLog lt_flush]

### ISSUE THREAD

**Issue by qq510304723** [CocoaLumberjack/CocoaLumberjack#1021]
my app often crash on [DDLog lt_flush] , this is srack:
`
#23 Thread
NSFileHandleOperationException
*** -[NSConcreteFileHandle synchronizeFile]: Operation not supported

CoreFoundation | ___exceptionPreprocess + 124
-- | --
1 libobjc.A.dylib | objc_exception_throw + 56
2 CoreFoundation | -[NSException initWithCoder:]
3 Foundation | __NSFileHandleRaiseOperationException + 160
4 CCOA | __17-[DDLog lt_flush]_block_invoke (DDLog.m:891)
5 libdispatch.dylib | __dispatch_call_block_and_release + 24
6 libdispatch.dylib | __dispatch_client_callout + 16
7 libdispatch.dylib | __dispatch_queue_serial_drain + 240
8 libdispatch.dylib | __dispatch_queue_invoke + 652
9 libdispatch.dylib | __dispatch_queue_override_invoke + 360
10 libdispatch.dylib | __dispatch_root_queue_drain + 572
11 libdispatch.dylib | __dispatch_worker_thread3 + 124
12 libsystem_pthread.dylib | _pthread_wqthread + 1288
13 libsystem_pthread.dylib | start_wqthread + 4
`

i think maybe fileHandle is be nil when synchronizeFile in DDFileLogger,
```objective-c
- (void)dealloc {
    [_currentLogFileHandle synchronizeFile];
    [_currentLogFileHandle closeFile];

    if (_currentLogFileVnode) {
        dispatch_source_cancel(_currentLogFileVnode);
        _currentLogFileVnode = NULL;
    }

    if (_rollingTimer) {
        dispatch_source_cancel(_rollingTimer);
        _rollingTimer = NULL;
    }
}
```

when dealloc , i think _currentLogFileHandle should be set to nil , 

thanks!!!

**Comment by qq510304723** [comment]
<img width="688" alt="2019-01-22 11 14 52" src="https://user-images.githubusercontent.com/7485345/51510891-f2fac280-1e39-11e9-8919-8769f3a2f478.png">

**Comment by sushichop** [comment]
Thanks for reporting issue.
Can you fill the description about you environment? 

Then, I think this issue has been improved at master branch (probably).
If you are interested in it, could you try it?
Thanks again.

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / osx / tvos / watchos
 Platform Version        | e.g. 8.0
 CocoaLumberjack Version | e.g. 2.3.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 7.3
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Demo project link       | e.g. link to a demo project that highlights the issue

**Comment by hhanesand** [comment]
@sushichop Found this : 

```swift
internal class FileHandle {
	let fileDescriptor: Int32
	let pointer: UnsafeMutablePointer<FILE>

	init(fileDescriptor fd: Int32) {
		fileDescriptor = fd
		pointer = fdopen(fd, "r+")
	}

	func write(data: NSData) -> Result<()> {
		return Darwin.write(fileDescriptor, data.bytes, UInt(data.length)) != 0 ? success() : failure()
	}

	func flush() -> Result<()> {
		return fflush(pointer) != 0 ? success() : failure()
	}

	deinit {
		fclose(pointer)
	}
}
```

See https://github.com/Carthage/Carthage/issues/350

**Comment by sushichop** [comment]
@hhanesand Thanks for your information😃
I had a glance at https://github.com/Carthage/Carthage/issues/350 . It seems that `flush` is not used now and it was used for flushing  stdout/err. And I also have seen `flush` method was used for stdout/err in other projects(e.g. Server side Swift proj), but I'm not sure this method is useful to fix? our  current issue...

Then, I think current issue is related to using `flush` method that is public method in CocoaLumberjack. Or if people who use CocoaLumberjack 3.4.2 or before, (as you know )they will encounter app's crash because its versions had thread racing issue😉

So I think we need more environment info(e.g. CocoaLumberjack versions), as I said above.
Thanks.

**Comment by qq510304723** [comment]
> Thanks for reporting issue.
> Can you fill the description about your environment?
> 
> Then, I think this issue has been improved at master branch (probably).
> If you are interested in it, could you try it?
> Thanks again.
> 
> Info   |	Value
> -- | --
> Platform Name | iOS
> Platform Version | 11.2.6 (15D100)
> CocoaLumberjack Version | 3.4.2	
> Integration Method | 
> Xcode Version | 	Version 10.1 (10B61)
> Repro rate | 	
> Demo project link | [DemoAppRepo](https://github.com/qq510304723/DemoAppRepo.git)

Thanks for reply

**Comment by hhanesand** [comment]
Took a peak as well, and couldn't find anything wrong with your code, but I couldn't reproduce the crash either. `[DDLog lt_flush]` is only called by CL when `applicationWillTerminate` is called, so unless you're calling `[DDLog flushLog]` from somewhere else in your project, this should only crop up when the app is force quit. Have you ever observed this error while developing?

Only thing other thing is that you are importing both `DDLog.h` and `CocoaLumberjack.h`. This activates our legacy logging macros and our new ones (I think?) @CocoaLumberjack/collaborators anyone know if this could cause issues?

**Comment by sushichop** [comment]
> Only thing other thing is that you are importing both `DDLog.h` and `CocoaLumberjack.h`.

Good catch🙂
I'm not sure this causes crashing, but it is undesirable usage(I confirmed that legacy macros was actually imported).

**Comment by sushichop** [comment]
@qq510304723
`CocoaLubmerajck 3.5.2` has been released.
https://github.com/CocoaLumberjack/CocoaLumberjack/releases/tag/3.5.2

Can you try to use it and import `<CocoaLumberjack/CocoaLumberjack.h>` instead of importing `DDLog.h` directly. It's an easy how to use.

I'm going to close this issue for now as I haven't heard from you in a while unfortunately.
Feel free to reopen or create another issue if you still encounter a problem.

**Comment by qq510304723** [comment]
> @qq510304723
> Thanks for your information. I had a glance at your code, and your code itself seems to have no problem.
> 
> Then you need to update CocoaLumberjack to `3.5.0`. CocoaLumberjack `3.4.2` had several issues about DDFileLogger and now they have already been fixed by #946, #986, #1003, and etc.
> And if your iOS minimum version is 10.0 or later, it is recommended to use `DDOSLogger` instead of `DDASLogger`.

Thanks you, i will try upgrade and follow it
